### ISSUE TITLE
General Loss function discussion

### ISSUE THREAD

**Issue by KayARS** [stepankonev/waymo-motion-prediction-challenge-2022-multipath-plus-plus#9]
I had some general questions about the **nll_with_covariances** you use for your loss function.
As it seems there are a lot of other people that have similar questions right now on your repo.
Could you share with us, why you've chosen this function? Did you get it from somewhere, or did you do it by yourself?
Whats the general idea behind it? I'm seeing a lot of log and exponent calls, but I figure there is some mathematical formula behind this that I can't find anywhere.

The interval of the loss seems to be [-inf, inf], which seems a little bit weird to me, because it's hard to tell when the model is actually converging.
(I guess negative is good, but I can't say for sure)

Any help would be greatly appreciated!

**Comment by weihuaLi1** [comment]
I think the author of MultiPath++ use the loss defined in [MultiPath](https://arxiv.org/abs/1910.05449). And the author of this repo may use similar idea in the code.
But the operation "torch.logdet" may create 'nan' or 'inf' in code and raise error in the following code.

**Comment by KayARS** [comment]
Yes if seen that, too.
Did you find a good solution when the determinant of the covariance matrix is <=0 ? 
I'm currently just setting the determinant to a small positive value, (that the log doesn't return -inf or nan)
but I'm not sure if that is an optimal approach.

**Comment by weihuaLi1** [comment]
I haven't found a reasonable solution. @stepankonev , Could you please help us?

**Comment by KayARS** [comment]
I've been recommended to do following (which is generally helpful when using the log):
torch.log(torch.det(covariance_matrices).unsqueeze(-1)+1e-6)
instead of:
torch.logdet(covariance_matrices).unsqueeze(-1)

Might not be the most optimal, but it works for me and might help you, too.
It seems to be a general instability problem

**Comment by stepankonev** [comment]
Hi everyone! You can read about the loss function in the technical report, actually it is nothing but NLL that was used in many papers

**Comment by xushilin1** [comment]
Is there a classification loss in your code?

**Comment by xushilin1** [comment]
https://github.com/stepankonev/waymo-motion-prediction-challenge-2022-multipath-plus-plus/blob/4636641b00cd16f62f82a612c728200f3fc4a32b/code/model/losses.py#L15

I want to know why there is a log_softmax(confidences) instead of classification loss.

**Comment by Xinchengzelin** [comment]
> I've been recommended to do following (which is generally helpful when using the log): `torch.log(torch.det(covariance_matrices).unsqueeze(-1)+1e-6)` instead of: `torch.logdet(covariance_matrices).unsqueeze(-1)`
> 
> Might not be the most optimal, but it works for me and might help you, too. It seems to be a general instability problem


> I've been recommended to do following (which is generally helpful when using the log): `torch.log(torch.det(covariance_matrices).unsqueeze(-1)+1e-6)` instead of: `torch.logdet(covariance_matrices).unsqueeze(-1)`
> 
> Might not be the most optimal, but it works for me and might help you, too. It seems to be a general instability problem

I find [here](https://math.stackexchange.com/questions/889425/what-does-determinant-of-covariance-matrix-give/889513#889513?newreg=1e6ce152e0ce4c25bc8d7a6cf3e2e6fc%20), covariance matrix is  semi-definite, `torch.det(covariance matrix)` >= 0

Could you help me explain what this mean? 
```
errors = coordinates_delta.permute(0, 1, 2, 4, 3) @ precision_matrices @ coordinates_delta
```

**Comment by Xinchengzelin** [comment]
I find the [slides](https://www.cs.ubc.ca/~schmidtm/Courses/540-W18/L15.pdf  ), maybe it's helpful
